eupv: Don't error out on missing flatpaks

mwleeds opened the pull request:
In a perfect world, USB drives created with eos-updater-prepare-volume
could be used to update Endless computers on any previous OS versions,
but alas, this is not a perfect world. One of the reasons we can't keep
that promise is checkpoints ("eos.checkpoint-target"), but we also can't
guarantee we'll have all the flatpaks needed by
eos-updater-flatpak-installer on the target machine receiving the
update. This is because flatpaks may be uninstalled by the user after
being auto-installed, or they may be installed and later uninstalled by
eos-updater-flatpak-installer, as will be the case for
com.hack_computer.ProjectLibrary in 3.9.1.

So make failures of the `flatpak create-usb` command non-fatal for
flatpaks not explicitly specified by the user. We may also want to
install any missing flatpaks before doing the copy when we have an
Internet connection, but that improvement is left as an exercise for the
reader.

https://phabricator.endlessm.com/T30368

wjt commented on eos-updater-prepare-volume/eos-updater-prepare-volume:
```diff
@@ -56,11 +56,17 @@ class VolumePreparer:
+        try:
+            subprocess.check_call(cmd)
+        except subprocess.CalledProcessError:
+            if (fatal_error):
```
```suggestion
            if fatal_error:
```

mwleeds replied:
I'm too used to C code :laughing:

wjt reviewed (changes requested):
If I understand correctly, this will allow the OS commit D on the USB stick to be used to update a machine from some previous commit C provided that the autoinstall file changes between C and D do not require any of the missing Flatpaks to be installed. But if C → D requires that one of the missing Flatpaks be installed, that update will still fail. Right?

So, this seems like an improvement, but it's another nail in the box of "we should make e-u-f-i failures non-fatal" IMO.

In the specific example you hit: ProjectLibrary was previously installed as part of an old update, and is now removed in 3.9.1. So this is a non-issue: it's missing, but 3.9.1 requires it be removed, so that's OK. If we had removed the old stanza that installed it from the autoinstall, would this change have been necessary? Also, does `_get_autoinstall_flatpaks()` ignore install actions which are subsequently countermanded by an uninstall action?

mwleeds commented:
> If I understand correctly, this will allow the OS commit D on the USB stick to be used to update a machine from some previous commit C provided that the autoinstall file changes between C and D do not require any of the missing Flatpaks to be installed. But if C → D requires that one of the missing Flatpaks be installed, that update will still fail. Right?
> 
Right
> So, this seems like an improvement, but it's another nail in the box of "we should make e-u-f-i failures non-fatal" IMO.
> 
I would tend to agree
> In the specific example you hit: ProjectLibrary was previously installed as part of an old update, and is now removed in 3.9.1. So this is a non-issue: it's missing, but 3.9.1 requires it be removed, so that's OK. If we had removed the old stanza that installed it from the autoinstall, would this change have been necessary?

No, it wouldn't have been necessary for this case (though as you point out it's still a useful change for the case when something has been manually uninstalled).

>Also, does `_get_autoinstall_flatpaks()` ignore install actions which are subsequently countermanded by an uninstall action?

No, but that would be an easy change to make. I'll do that.

mwleeds commented:
There is an edge case that's quite hard to support: if there's a checkpoint commit at or before a flatpak is auto-installed, and that flatpak is later auto-uninstalled, the update USB won't be able to provide the flatpak for a receiving computer that needs to follow the checkpoint. However, the USB also won't have the checkpoint itself, so it's a broader problem than we can fix here.

Probably another reason to avoid checkpoints unless truly necessary, for now.

mwleeds commented:
Checked that this script still behaves as expected with the new minor changes

wjt commented on eos-updater-prepare-volume/eos-updater-prepare-volume:
```diff
@@ -150,6 +156,8 @@ class VolumePreparer:
                EosUpdaterUtil.FlatpakRemoteRefActionType.UNINSTALL:
                 autoinstalls.add(action.ref.ref.format_ref())
+            else:
+                autoinstalls.discard(action.ref.ref.format_ref())
```
I have never encountered `set.discard` before! I have always used:

```python3
                autoinstalls -= {action.ref.ref.format_ref()}
```

when I've wanted this, which has always felt clunky. Nice.

mwleeds replied:
Yeah just discovered that too

wjt commented on eos-updater-prepare-volume/eos-updater-prepare-volume:
```diff
@@ -208,9 +209,18 @@ class VolumePreparer:
+        # For specified flatpaks, an error is fatal
+        for flatpak_ref in self.flatpak_refs:
+            self.__run(['flatpak', '--system', 'create-usb',
+                        self.volume_path] + [flatpak_ref],
```
```suggestion
                        self.volume_path, flatpak_ref],
```

wjt commented on eos-updater-prepare-volume/eos-updater-prepare-volume:
```diff
@@ -208,9 +209,18 @@ class VolumePreparer:
+        for flatpak_ref in autoinstall_flatpaks:
             self.__run(['flatpak', '--system', 'create-usb',
-                        self.volume_path] + all_flatpak_refs)
+                        self.volume_path] + [flatpak_ref],
```
```suggestion
                        self.volume_path, flatpak_ref],
```

wjt reviewed (changes requested):
LGTM except for a pair of minor changes I missed before

mwleeds commented:
> > So, this seems like an improvement, but it's another nail in the box of "we should make e-u-f-i failures non-fatal" IMO.
> 
> I would tend to agree

Filed https://phabricator.endlessm.com/T31149
